### PR TITLE
Network protocol efficiency

### PR DESCRIPTION
**Pull request by ciribob**
- Added UPNP to the server
- Implemented JF17 new and old interfaces
- Changed Protocol + Sync to only need to be told if there is a change, no heartbeat of UPDATE (with coalition required)
- Removed unneeded fields for DCS sync and Server TCP Sync
- UPDATE messages only sent with a new position if Line of sight or distance is enabled on the server
- Radio update only sent on changes

### CODE REVIEW

**Comment by ciribob** [comment]
@rkusa 
@rurounijones 
@ace747 

As above - no actual changes to the messages themselves being sent - other than the removal of unneeded and null fields - but clients dont require an update message for another transmission to be valid

The minimum you need to send is one RADIO_UPDATE message containing

Coalition
Name
RadioInfo
Position (can be just 0 for position)

and thats it

Once you've sent a fully formed RADIO_UPDATE no further heartbeats or updates need to be sent on the TCP interface

No changes at all to UDP so all existing code will work fine

You should see a huge drop in traffic as the 5 second heartbeat sync and 30 seconds radio update to servers is no longer sent by standard clients

**Comment by rkusa** [comment]
Thanks for the heads-up 👍

**Comment by ciribob** [comment]
I'll post an update later today so you have a beta to test against. Any questions let me know but hopefully just means removing a bunch of code! 👍

**Comment by ciribob** [comment]
Update here: https://github.com/ciribob/DCS-SimpleRadioStandalone/releases/tag/1.8.0.0

**Comment by rkusa** [comment]
> A client is determined as "not in game" if all radios are set to Modulation 3 which is disabled. Just set one of the 10 radios in RadioInfo.radios to not modulation 3

Is this what a client should do when disconnecting? E.g. in terms of the ATIS stations, should I send an update with all radios disabled once I stop an ATIS station (and the ATIS client should be considered as not connected anymore)?

**Comment by ciribob** [comment]
> > A client is determined as "not in game" if all radios are set to Modulation 3 which is disabled. Just set one of the 10 radios in RadioInfo.radios to not modulation 3
> 
> Is this what a client should do when disconnecting? E.g. in terms of the ATIS stations, should I send an update with all radios disabled once I stop an ATIS station (and the ATIS client should be considered as not connected anymore)?

I've just been going through this again - and I'm going to remove the distinction between in game and not, just going with straight connected or not. It can always come back later but going for the minimum of sync data and i'll build up again I think.

For ATIS - an SRS Receiving client doesn't actually check the sending clients RadioInfo as all the information is in the UDP packet so if you just send one RADIO_UPDATE to initialise the client on the server (and all other connected clients) that'll be all you need to do. 

Going against what I said earlier, leave all 10 radios set to disabled on their modulation. This means you wont receive transmissions sent on any frequency over UDP as they're not required for you and technically you're not in game anyway

Better use case is enabling the Show Tuned Clients setting as that actually tells you who can hear you, rather than could maybe hear you if you found the right frequency

Minimum JSON to register with the server (all radios disabled) with a single RADIO_UPDATE
`{
   "Client":{
      "ClientGuid":"_hITenjPI0GhX4-5ISHgYA",
      "Name":"New callsign",
      "Coalition":2,
      "RadioInfo":{
         "radios":[
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            }
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            }
            {
               "enc":false,
               "encKey":1,
               "freq":30000000.0,
               "modulation":1,
               "secFreq":1.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            },
            {
               "enc":false,
               "encKey":0,
               "freq":1.0,
               "modulation":3,
               "secFreq":0.0
            }
         ],
         "unit":"External AWACS",
         "unitId":200000001,
         "iff":{
            "control":2,
            "mode1":-1,
            "mode3":-1,
            "mode4":false,
            "mic":-1,
            "status":0
         }
      },
      "LatLngPosition":{
         "lat":0.0,
         "lng":0.0,
         "alt":0.0
      }
   },
   "MsgType":3,
   "Version":"1.8.0.0"
}`

As above this will mean no client tries to do line of sight calculations as Position is all 0s
LotATC will be happy as IFF/ transponder is set to disabled
SRS Server will not transmit any VoIP UDP to you as you have no receiving radios

Hope that makes sense!

**Comment by ciribob** [comment]
Binaries updated on the 1.8.0.0. or use the server below to test against

Should be the final set of changes networking wise 👍 

[SR-Server.zip](https://github.com/ciribob/DCS-SimpleRadioStandalone/files/4535258/SR-Server.zip)

**Comment by rkusa** [comment]
DATIS received [an update](https://github.com/rkusa/DATIS/releases/tag/1.0.0-alpha.9) that works with the changes, thanks again for the explanations 👍

**Comment by ciribob** [comment]
@rkusa glad to hear! I'll work on getting a release out on the weekend

File: DCS-SR-Client/Network/SRSClientSyncHandler.cs [hunk 1]
```diff
@@ -302,8 +288,11 @@ private void ClientSyncLoop()
                                     case NetworkMessage.MessageType.RADIO_UPDATE:
                                     case NetworkMessage.MessageType.UPDATE:
 
-                                        _serverSettings.Decode(serverMessage.ServerSettings);
-
+                                        if (serverMessage.ServerSettings != null)
```
Comment (ciribob) [inline DCS-SR-Client/Network/SRSClientSyncHandler.cs:291]: @rurounijones

**Comment by rurounijones** [comment]
[Overlord Bot](https://github.com/rurounijones/DCS-SimpleRadioStandalone) has been updated to 1.8

**Comment by ciribob** [comment]
@rkusa shouldn't affect you but for @rurounijones it will

I've added extra modulation types for some of the future modules including the Hornet MIDS and A10C HAVEQUICK

>  /*
>         AM = 0,
>         FM = 1,
>         INTERCOM = 2,
>         DISABLED = 3,
>         HAVEQUICK = 4, //unsupported currently
>         SATCOM = 5, //unsupported currently
>         MIDS = 6,*/

MIDS works by picking a frequency based off the currently selected channel

so channel 126 is actually: 1030 MHz + (126 *0.1MHz) 

Real system is digital but this is fine for our purposes

For @rurounijones should just be another pull from this branch

I'll finish off the MIDS feature on this branch, probably tonight and release Saturday or Sunday

**Pull request merged**